### DISCUSSION TITLE
OptimizeBroadPhase usage

### DISCUSSION THREAD

**Discussion by turanszkij** [jrouwe/JoltPhysics#1226] [General]
Hi, I'm currently running OptimizeBroadPhase every frame on a background thread - just in case - and I wonder if there could be a problem with my approach in theory (I don't notice a problem in practice currently):

 1) The OptimizeBroadPhase always starts on a background thread after PhysicsSystem::Update was called for the last time in the current frame.
 2) Meanwhile, I update my game objects from the Jolt bodies, which are accessed through GetBodyInterfaceNoLock()
 3) During the rest of the frame, it's also possible that constraints will be created infrequently or forces, etc. applied on bodies. I also use GetBodyInterfaceNoLock() for these, the PhysicsSystem::Update() is guaranteed to not be updating here, but the OptimizeBroadPhase() could still be running
 4) Before the next frame PhysicsSystem::Update(), the thread with OptimizeBroadPhase() is guaranteed to be finished.

Point 3) is the most worrying since that could be modifying the physics world while OptimizeBroadPhase() hasn't finished. Does this sound like could lead to threading issues?

**Comment by jrouwe** [comment]
Hello,

There's no point in calling `OptimizeBroadPhase` every frame. `PhysicsSystem::Update` does the same thing but only does part of the work every frame. `OptimizeBroadPhase` is only there to be called the first frame after loading so that you don't start the game with a really inefficient broadphase (which will then take 10 `PhysicsSystem::Update`'s or so to incrementally fix).

If you call `OptimizeBroadphase` and use `GetBodyInterfaceNoLock` from another thread, you could have a race condition where you're modifying the bounding box of a body while it is being processed in the broadphase.

**Reply by Mahmoud1205** [reply]
If I call OptimizeBroadphase the first frame, then after a few seconds add a lot of new physics bodies in one frame, should I call OptimizeBroadphase again?

**Reply by jrouwe** [reply]
It doesn't hurt (other than that it could cause a frame rate hiccup). As long as you've batch added all those new bodies using `BodyInterface::AddBodiesPrepare` / `AddBodiesFinalize` and they're somewhat separated in space from the existing bodies (e.g. it's a new level section) then you shouldn't really need to call `OptimizeBroadPhase`. The best way to find out is to profile it, if the physics update is significantly faster after `OptimizeBroadPhase` then it could be worth it.

**Reply by Mahmoud1205** [reply]
Okay! Thanks for your continuous help

**Comment by turanszkij** [comment]
Thank you, I will rewrite this part.
